daemon: Add socket activation via /run/rpm-ostreed.socket

cgwalters opened the pull request:
This is a re-submit of https://github.com/coreos/rpm-ostree/pull/2932

---

For historical reasons, the daemon ends up doing a lot of
initialization before even claiming the DBus name.  For example,
it calls `ostree_sysroot_load()`.  We also end up scanning
the RPM database, and actually parse all the GPG keys
in `/etc/pki/rpm-gpg` etc.

This causes two related problems:

- By doing all this work before claiming the bus name, we
  race against the (pretty low) DBus service timeout of 25s.
- If something hard fails at initialization, the client can't
  easily see the error because it just appears in the systemd
  journal.  The client will just see a service timeout.

The solution to this is to adopt systemd socket activation,
which drops out DBus as an intermediary.  On daemon startup,
we now do the process-global initialization (like ostree
sysroot) and if that fails, the daemon just sticks around
(but without claiming the bus name), ready to return the
error message to each client.

After this patch:

```
$ systemctl stop rpm-ostreed
$ umount /boot
$ rpm-ostree status
error: Couldn't start daemon: Error setting up sysroot: loading sysroot: Unexpected state: /run/ostree-booted found, but no /boot/loader directory
```

Thank you for contributing to rpm-ostree.

If you are adding functionality to tree composes, please add
a corresponding test to the compose-test suite. Similarly,
if adding a client-facing feature, consider the vmcheck
suite. Regressions fixes are also great candidates for new
tests.

If you're not sure where or how to add tests, don't hesitate
to ask for help from the maintainers.

Cheers!

cgwalters commented:
This blocks on SELinux policy updates: https://bugzilla.redhat.com/show_bug.cgi?id=2110012

cgwalters commented on rust/src/daemon.rs:
```diff
@@ -131,6 +132,108 @@ fn deployment_populate_variant_origin(
+    // On success, we spawn a helper thread that just responds with
+    // sucess to clients that connect via the socket.  In the future,
+    // perhaps we'll expose an API here.
+    std::thread::spawn(move || process_clients(listener, &Ok(())));
```
We can use tokio now particularly after https://github.com/coreos/rpm-ostree/pull/3575

cgwalters commented on src/daemon/rpm-ostreed.socket:
```diff
@@ -0,0 +1,9 @@
+
+[Socket]
+ListenStream=/run/rpm-ostree/client.sock
+SocketMode=0600
```
Need to change this to be 0666 *or* make the connections to it only run if we're root.

Also xref https://github.com/coreos/rpm-ostree/issues/3850

cgwalters commented:
OK, I now made this into a build-time option.  (Though looking at it, I also need to make installation of the socket unit a build-time option)

This will allow us to ship this in F37 now (with a matching specfile conditional) and gain experience with it.

Leaving as draft though for now since this is a nontrivial change and could use some more design thought before merging.

cgwalters commented:
/test all

cgwalters commented:
Dang it, I did a whole port to https://crates.io/crates/uds before discovering https://github.com/tormol/uds/issues/6

For now...I switched back to using worker threads.  That said...hmmm.  What would probably make sense here is to factor out an `async-sock-seqpacket` crate...we've now got very similar code in all 3 of:

- https://github.com/coreos/bootupd/blob/ff9555afad3120eade3d7d4611eecd8391ae21fc/src/ipc.rs#L46
- https://github.com/containers/containers-image-proxy-rs/blob/a4223a034895bf27423878f3af2ba22d50fc7b36/src/imageproxy.rs#L96
- this project

I guess one can tell that I like `SOCK_SEQPACKET`?

cgwalters commented:
If we want to support a non-DBus API, one thing we'll likely need to do for now is bridge requests from the tokio thread to the glib/DBus daemon thread.  We can do this by acquiring that worker thread's main context and then invoking https://docs.rs/glib/latest/glib/struct.MainContext.html#method.spawn

cgwalters commented:
OK C9S and RHEL9.1 should include the necessary selinux policy changes for this, so I think this can be considered for review and landing.

Looks like there's a CI failure in the Jenkins/vmcheck bits I need to dig into.

But right now more problematically the socket path isn't being tested by CI at all; need to add a f37 build or extend our c9s build to enable this feature and actually do tests too.

jmarrero reviewed (approved):
lgtm

cgwalters commented:
I've been looking in the background at https://github.com/mitsuhiko/tokio-unix-ipc/ and have [two issues open](https://github.com/mitsuhiko/tokio-unix-ipc/issues/created_by/cgwalters), though neither are blocking.

RishabhSaini commented on src/daemon/rpm-ostreed.socket:
```diff
@@ -2,7 +2,7 @@
 
 [Socket]
-ListenSequentialPacket=/run/rpm-ostree/client.sock
+ListenStream=/run/rpm-ostree/client.sock
```
We are declaring a `Sequential Packet` type socket in daemon.rs and client.rs, but declaring it as an `Stream` type in this WIP commit when adding the tokio-unix-ipc. For our usecase does it make a difference?

cgwalters replied:
I think we need to use `ListenSequentialPacket` if we're doing https://github.com/coreos/rpm-ostree/pull/3874/files#diff-bc43d479a8bbd812cda2b223dee82b4540dbd1ee93200a784301a55fc1cfe905R171

cgwalters commented on Cargo.toml:
```diff
@@ -94,6 +94,7 @@ tracing-subscriber = { version = "0.3", features = ["env-filter"] }
 tokio = { version = "1.28.2", features = ["time", "process", "rt", "net"] }
 xmlrpc = "0.15.1"
 termcolor = "1.1.3"
+tokio-unix-ipc = { version = "0.2.0", features = ["serde"] }
```
It looks like I didn't start using it in this version of the PR; we are just serializing JSON over SOCK_SEQPACKET.

The advantage of tokio-unix-ipc is that it actually integrates well into async rust, unlike `SOCK_SEQPACKET` today.

But we could probably proceed without using `tokio-unix-ipc` to start.

RishabhSaini commented on rust/src/client.rs:
```diff
@@ -198,6 +243,18 @@ pub(crate) fn client_start_daemon() -> CxxResult<()> {
+    #[cfg(feature = "client-socket")]
+    return start_daemon_via_socket().map_err(Into::into);
+    #[cfg(not(feature = "client-socket"))]
+    return start_daemon_via_systemctl().map_err(Into::into);
```
Should we completely remove this? Since we are assuming that the feature of client-socket exists by default?

cgwalters replied:
Yes, we should safely be able to remove that code!

RishabhSaini commented on rust/src/daemon.rs:
```diff
@@ -131,6 +149,167 @@ fn deployment_populate_variant_origin(
+            // with a socket.
+            init_res.context("Initializing via dbus")?;
+            tracing::debug!("Initializing directly (not socket activated)");
+            let listener = cfg!(feature = "client-socket")
```
The `None` case would not exist if we can no longer start the daemon from dbus and only activated via the socket

cgwalters replied:
Ahh...this is a complicated topic!  We definitely must continue to support clients that aren't `rpm-ostree` starting the daemon via DBus.

So hmm...two ways to do that.

One is just having an integration test that's like
```
systemctl stop rpm-ostreed
rpmostree_busctl_call_os ...
```

RishabhSaini replied:
Or could it?

RishabhSaini replied:
yes this fails since when the daemon is `idle` or `stopped`, and `systemctl start` or `busctl` is invoked. We may need a more graceful handling of re-connecting to the socket.

RishabhSaini commented on rust/src/client.rs:
```diff
@@ -156,6 +156,47 @@ pub(crate) fn client_handle_fd_argument(
+fn start_daemon_via_socket() -> Result<()> {
+    use cap_std::io_lifetimes::IntoSocketlike;
+
+    let conn = tokio::net::UnixStream::connect("/run/rpm-ostree/client.sock")?;
```
`tokio::net::UnixStream::connect("/run/rpm-ostree/client.sock").await?` will turn this function into an async

cgwalters replied:
Right, we can probably just make it synchronous and use `std` here.

RishabhSaini replied:
Trying to `connect` after the daemon is idle or closed results in a `os error 111: connection refused`. This is because the daemon is no longer listening on the socket

RishabhSaini commented on rust/src/daemon.rs:
```diff
@@ -131,6 +149,167 @@ fn deployment_populate_variant_origin(
+                    rustix::net::Protocol::from_raw(0),
+                )?;
+                let addr = crate::client::sockaddr()?;
+                rustix::net::sockopt::set_socket_reuseaddr(&socket, true)?;
```
Upon a `start-daemon` invocation, the dameon gives a `error: Address already in use (os error 98)` . This happens when we try bind the socket to the address despite the `so-reuseaddr` option enabled

cgwalters replied:
OK right, we may need to drop the code that tries to bind to the socket in the daemon code.  

Hmm...I think what we really want is to make systemd *always* pass the socket to us even when we're started via dbus.

I think we may be able to do this via something like
```
[Service]
StandardInput=fd:rpm-ostreed.socket
```
perhaps?

Then we should always be able to find our socket on stdin, if that works.  Needs testing

cgwalters replied:
OK that didn't work in my testing, but using `Sockets=rpm-ostreed.socket` did seem to work.

I was actually testing with this:

```
[root@cosa-devsh ~]# cat /etc/systemd/system/foo.socket 
[Socket]
ListenStream=/run/foo.sock
SocketMode=0600

[Install]
WantedBy=sockets.target
[root@cosa-devsh ~]# cat /etc/systemd/system/foo.service 
[Service]
Sockets=foo.socket
Type=oneshot
ExecStart=env
ExecStart=ls -al /proc/self/fd/0
[root@cosa-devsh ~]# 
```

Using `Sockets=` I do see when I `systemctl start foo` that it has the expected `LISTEN_FDS=1` in the environment.

cgwalters replied:
For the record, I seem to have just been wrong here, I do seem to get the socket always passed.

I suspect the problem we have been seeing here is in the case where the socket unit isn't active.  That will happen in some development scenarios (e.g. copying new rpm-ostree binary, but not the systemd unit - or even more commonly, forgetting to systemctl daemon-reload etc.)

Arguably perhaps, the daemon should just yell loudly about this and the client can also loudly complain if the socket is missing, and just fall back to dbus.

cgwalters commented:
OK I made a few more changes here just to clean things up:

- Squashed all the commits
- Removed now-spurious whitespace changes leftover from the removed feature flag
- Changed you to be the author, listed me as co-author

cgwalters commented on rust/src/daemon.rs:
```diff
@@ -131,6 +149,139 @@ fn deployment_populate_variant_origin(
     Ok(())
 }
 
+pub(crate) fn write_message(conn: &OwnedFd, message: SocketMessage) -> Result<()> {
```
I think now that we're using a stream, we probably need to change all this so that we write exactly one message and then close the connection.  And relatedly, instead of using `send` we need to `write_all`.

RishabhSaini replied:
Does this mean we send a `rustix::net::shutdown` from the daemon after we have done acknowledgment?

cgwalters commented on rust/src/daemon.rs:
```diff
@@ -131,6 +149,139 @@ fn deployment_populate_variant_origin(
+    match init_res {
+        Ok(()) => {
+            dbg!(&listener);
+            std::thread::spawn(move || run_acknowledgement_worker(listener));
```
Also now that we're using a streaming protocol we can switch all of this to use a proper async tokio task instead of a thread.  But, not a blocker.

RishabhSaini replied:
so instead of spawning thread to run task concurrenty we use `tokio::task::spawn(async move {run_acknowledgement_worker(listener).await })`. But wouldn't this require the `daemon_main` to be async or do we want to block on this with `futures::executor::block_on(async {...})`

RishabhSaini commented:
The jenkins pipeline shows `error: cleanup: Loading sysroot: Starting daemon via socket: Connecting: No such file or directory (os error 2)`. Does the `rpm-ostreed.socket` file in `/etc/systemd/system` need to be enabled?

cgwalters commented:
I think what's going on here is:

We include the .socket unit in the RPM, but it's not enabled by default because it's not in [the preset list](https://src.fedoraproject.org/rpms/fedora-release/blob/rawhide/f/90-default.preset).

This has burned us before.  In theory, what we'd need to do is add it to that list.  However doing so will dramatically slow down our ability to ship this.

So what I'd propose is:

- Change the client code to check if it's running as root, and if so `systemctl start rpm-ostreed.socket` explicitly
- Continue to improve and land this PR
- Once it lands, add a PR to fedora-release to enable the socket by default
- Eventually, drop the client auto-starting

Hmm I think we may also need to add `%systemd_post` for our units.

cgwalters commented on rust/src/daemon.rs:
```diff
@@ -131,6 +144,146 @@ fn deployment_populate_variant_origin(
+    Ok(())
+}
+
+static SHUTDOWN_SIGNAL: Lazy<Mutex<Option<Sender<()>>>> = Lazy::new(|| Mutex::new(None));
```
I was looking at this again and I remembered that there's a race condition this was intended to fix but we didn't.  Basically, the socket clients aren't accounted for in our current exit-on-idle logic.

What we need to do is:

- If we have a socket client, then disable the idle exit
- When we choose to idle exit, first shut down the `accept()` handler

cgwalters replied:
Mmm and a further complication here is to make this race-free we probably need to use the glib main loop (aka main thread) and not the tokio thread - at least to *accept* clients on the socket; we could still do the actual processing on the tokio thread with appropriate synchronization.
